### ISSUE TITLE
WE NEED A GIF KEYBOARD TOOOooo!!11!1!!

### ISSUE THREAD

**Issue by matrixbot** [vector-im/element-meta#321]
Created by @ matthew:matrix.org.

**Comment by pixlwave** [comment]
Having this same conversation with someone right now. Their words not mine:
> Also, it won't be mainstream until it has gifs
> gifs are everything

🤔

**Comment by torinreilly** [comment]
I personally do not think there should be a _default_ gif search included in Matrix (or Element for that matter) that ties you to a specific GIF API or service. Instead, I think this is a good candidate for a widget (or widgets) so that each user can decide whether they want to opt in. 

I have been recently migrating myself and my friends over to Matrix as the new spot for our group chats. While not all of them are technical, most can appreciate the principle of self-hosting an e2e chat server and having more ownership of their data. What they **DID NOT** appreciate is the lack of a proper GIF keyboard.

As a result, I have been working on the following gif keyboard for Matrix and testing it amongst my friends: [torinreilly/matrix-gif-keyboard](https://github.com/torinreilly/matrix-gif-keyboard). 

It works by overriding the sticker-picker widget and sending the animated gifs as stickers. You can host the server yourself and it can pull gifs from Tenor, GIPHY, or from a custom list of sites using DuckDuckGo (not sure if this is worth it as discussed [here](https://github.com/torinreilly/matrix-gif-keyboard/issues/7)).

**Here's the catch.** The way stickers are implemented in the Element clients means the sticker URL has to be an MXC URI. In other Matrix clients like FluffyChat, stickers can have remote URLs (unfortunately FluffyChat does not support widgets though :/). So prior to sending the GIF it must be uploaded to a Matrix server. I am accomplishing by providing the GIF keyboard app server a Matrix access token to perform the uploads. This is fast enough to be fairly imperceptible by the end user, but it also exposes an endpoint that can be used to upload just about any image to your Matrix content repo! Not good! The only saving grace is the rate limiting that is supported by the content repository `/upload` endpoint.

Outside of requiring a login to the widget, I can't think of a good way to ensure this endpoint isn't abused. Currently, there only a few values that can be passed to the widget including username and the matrix server url, but nothing that can be used to ensure the user is authenticated. Passing things like the username and matrix server url could be a bit of security through obscurity, but it is easily faked so not worth implementing IMO. Looking for suggestions [here](https://github.com/torinreilly/matrix-gif-keyboard/issues/4).

If we could get the Element clients to support stickers with remote URLs it would remove the need for and upload in the first place, while saving us disk space and removing the auth headaches. It is also possible that we could get this to work using other room events, but I ran into a lot of trouble trying to send most type of events from a widget (m.room.message including images, etc). If anyone has experience/advice along those lines I would love to pick your brain.

In the meantime, please check it out if you are interested! I personally would not implement this in any kind of production for the time being. If you would like to try it out and don't want to host it yourself, I could let you use one of my hosted GIF keyboard widgets but only if you promise to not abuse my server too much!

Here are the key **issues** I'd love some feedback on:
[A better way to authenticate uploads to the content repository](https://github.com/torinreilly/matrix-gif-keyboard/issues/4)
[Does not work as a custom widget/integration](https://github.com/torinreilly/matrix-gif-keyboard/issues/1)
[Investigate if maintaining the DuckDuckGo API backend is worth it](https://github.com/torinreilly/matrix-gif-keyboard/issues/7)

Here is an example of how it looks on Element on iOS:
<img src="https://user-images.githubusercontent.com/67698871/151095384-8109e667-671d-4ff4-b96e-128dbb78c9cd.png" alt="iOS screenshot" width="200">

**Comment by SethFalco** [comment]
(Note, I'm not a maintainer, or even contributor for that matter.)

@torinreilly While I can appreciate your comment and approach to this, I think it's disregarding user-experience, no?

There are users who don't even understand what a file system is, or think their documents were deleted because it's no longer in their "Recent" in Microsoft Office anymore.

A lot of my friends are non-technical, they don't want to tinker with things or "learn the software". If it's not immediately obvious, or not intuitive, then it might as well not be a feature at all to them. With GIF keyboards kind of becoming a "standard" feature, I don't think it's ideal to make extra steps to access them.

I think it'd be reasonable to implement a GIF server natively, it's up to the user if they choose to use it or not. An option to turn off the UI for it would be great, or even better, a way to provide custom sources. I don't think the default experience should be through a widget though, which most people new to Element won't even understand immediately.

That being said, your project and widget look awesome. I especially love that you've considered solutions beyond Giphy and Tenor. A nice source to also consider could be [Creative Commons Search](https://wordpress.org/openverse/)?

**Comment by torinreilly** [comment]
@SethFalco Hi, Seth. Thanks for the thoughtful response. I think you are right on a lot of points here and a flexible native solution COULD be ideal provided it allowed enough choice to the end user. The thing I find nice about widgets is you are not beholden to the design decisions of anyone! If someone doesn't like how my widget is implemented or which services it targets they could use a different one. Or even fork it themselves and make it available to whoever they like.

Just to clarify, my intent is not for everyone to set up their own GIF keyboard server. In my case, I have a single app server that uses a bot account's token to perform the uploads to the content repository, but the message sent is actually on the requesting users behalf. Provided someone has setup a server, each user only needs to complete the one time setup of adding the widget in place of the sticker-picker described here: https://github.com/torinreilly/matrix-gif-keyboard#3-enable-widget.There is also the ability to add it through the UI as a "Custom Integration", which only requires the URL the app is hosted at (still working out some issues with this route for my app, it displays fine but messages are not sent).

I agree that extra step is inconvenient and might be confusing to some users, but I think most would not balk at it provided there was a sufficient guide.

I will definitely look at adding more search options. The DuckDuckGo API does allow for filtering results by license, so I was planning to implement that BUT I think Creative Commons Search would probably provide better results in the case of more open licenses! Thank you!

**Comment by captainperth** [comment]
I still can't believe we don't have a solution for this issue. Must I continue to manually download and upload gif files to truly express myself in element? Will someone please think of the memes!

**Comment by arkitoure** [comment]
I've just made the move from Discord/Telegram for our org over to Matrix/Element and while doing customisations we just realised using gifs 'natively' is not available.

This has created a lot of frustration for our users as one would expect this day and age and are looking for a solution.

The Giphy bot solution is unintuitive as users simply don't know what to search for to get an exact match of what they wish to express. Moreover, they often don't remember the command to trigger the search.

Naturally, we would want to see a gif icon next to the emoji and attachment icons in the UI which triggers a selector and search field.

I like the @torinreilly solution as it deals with the encrypted room issue that the giphy bot does not. So just on first glance I would be inclined to have this baked into the UX as mentioned.

This is a **critical** feature that cant be missing in todays world of chat/messaging apps. I will be looking at the code to see how I can contribute.

@torinreilly were there any security issues or like thoughts you had in implementing this solution? 


![Screenshot 2022-03-06 at 09 52 03](https://user-images.githubusercontent.com/7852460/156916049-999fd295-47bf-4384-a6dd-4f5652594fee.png)

**Comment by torinreilly** [comment]
@arkitoure There are definitely some security issues in my current solution. The way widgets are currently handled, there is no way to verify that the user of a widget is logged in to a specific Matrix server. In reality, there are only a handful of properties that are accessible to widgets (room name, username, etc) but any one of those could easily be faked in an http request so they really only provide security through obscurity. Any additional params you wish to include must be passed as part of the URL used when setting up the widget. 

In my current implementation, the widget app server has the ability to upload files directly to the Matrix server for which it is configured. This essentially gives anyone with access to the widget server URL cart-blanche to upload files to the Matrix server (not good!). 

This may be solved in a couple of ways: 

First, I am hoping that the often discussed support for using a Matrix server as an OpenID source comes to fruition. If this happens we could easily verify the user via OpenID. 

The second idea I had was to create a token or tokens that could be supplied as part of the widget URL during setup (ie https://widgeturl.com/?token=hdwry823hwekhr). This token could then be checked against a list of valid tokens by widget app server. This is better, but still relies on trusting users to not widely distribute tokens (might work OK for small communities or groups of friends) but for any large scale implementation the overhead of monitoring and revoking tokens used for abuse might not be worth it. (Additionally, I do not love the idea of passing the token as a URL param, but this is currently the only viable way to pass data to widgets as far as I know.)

The third option is to remove automatic upload to the server and require the user to copy and paste the GIF from the keyboard. I find this option to be less than intuitive and does not always work as expected, especially on mobile. It does however remove almost all of the risk of abuse by third parties.

The latter two solutions are less than ideal, but I could certainly add support for both of these with relative ease if there was demand.

**Comment by turt2live** [comment]
There are early designs for this up at https://www.figma.com/file/ifIvZA4XIP1JTyndHFr5Ub/Community-PRs?node-id=21%3A15691 however the technical undertaking of the proposed changes is a bit on the difficult side. 

Specifically, it relies on stickers being an API surface we can call due to them being a widget - this isn't impossible, and is in fact planned under my widgets work umbrella, but not something we have today. The designs also appear to rely on power levels for this stuff, which is a bit difficult to enforce at the moment, but not impossible with the current structure and future capabilities like extensible events.

For a code review perspective: I don't think we can reasonably accept PRs on this until product *and* design have taken a much closer look, possibly with an angle of custom stickers/emoji too, as otherwise the code is likely to bitrot or change dramatically.

**Comment by seanfarley** [comment]
Why is this a server api at all? Just make a client-only keyboard that pastes a gif into the chat. I want to like matrix but y'all have a damn-near infinite feature matrix that makes progress nigh impossible.

**Comment by turt2live** [comment]
For simple support, sure, we could do it entirely client side. For future options of potentially paid sticker packs and similar though, there needs to be a server :)

**Comment by seanfarley** [comment]
> For simple support, sure, we could do it entirely client side.

Then do that? Don't let perfect be the enemy of good.

> For future options of potentially paid sticker packs and similar though, there needs to be a server :)

Dear god, stop.

**Comment by michelebugio** [comment]
I feel like there should be a workable solution for the very short term, as the simple gif search and paste has now become such a widespread colloquialism that its absence prevents wider adoption of the whole matrix platform.

(it might seem like an exaggeration, but it's not, and it's also not the focus point, the focus here is to be able to search & paste gifs with the same ease as other messaging platforms).

A better structure with a server for the long term is perfectly fine, but it's not needed now.

**Comment by turt2live** [comment]
this is why I've kept it in the design team's queue and not assumed their previous designs would work: we might not have the engineering bandwidth to fix stickers, but we may very well have the bandwidth to add gif searching. 

For clarity: I was pointing out problems with the current designs, not proposing that the design team double down on the thing they've produced.

**Comment by ShadowJonathan** [comment]
I'm not sure why stickers are relevant in this issue? Imo, this issue explicitly (and only) calls for something akin to the `/giphy` command on discord, and the `@gif` bot on telegram, to have a simple lookup-and-paste for GIFs.

This can be achieved by client-only changes, while I have opinions on making stickers paid, I don't see how stickers are relevant at all to this issue.

**Comment by turt2live** [comment]
stickers and gifs are effectively the same thing at a technical level, regardless of presentation. As with most things Matrix, we don't generally want to tie ourselves to a specific gif provider so we'd probably lean on the sticker infrastructure a bit to anonymize access and generify the gif suppliers.

it's not a hard requirement.

**Comment by ShadowJonathan** [comment]
Alright, but then I just want to note that mentioning paid stickers brings in a very bitter point in the conversation that can further pollute and confuse it.

It was also framed in a way that stickers *has* to be a client-server api *because* it should become a paid offering, which strikes the wrong tone.

**Comment by turt2live** [comment]
it's no secret that stickers were always intended to grow a payments side to them - it would be entirely optional though. The challenge isn't the payment side though, it's that the current design mockups require gifs and stickers to be side-by-side with native UI: we'd need an API to be able to support this, which we don't have. Other options need to be explored, hence the call for re-design and product to be involved :/

**Comment by iceHtwoO** [comment]
How about the feature where a user can send the URL of a gif and the clients render it? This could be a compromise to make it a bit easier for the user and your not going to tie yourselfs to a particular provider.

**Comment by turt2live** [comment]
I think we generally want that for all images, yea. I can't find the issue easily though :( 

it's a similar idea to inline widgets where things like youtube URLs magically become embeds (optionally).

**Comment by moodler** [comment]
Travis, like many others, I believe this issue to be very important for widespread adoption of Matrix.  

I've had a conversation with Mathew Hodgson about it where he said the core team is busy and that "we welcome community contributions".  Which is totally fair enough.  

However, to now have issues like this effectively shut down with an over-complication and "the core team has to do it" at some possible unspecified time in the future is not great and counter to an open source ethos. 

How can we work as a community here to get a gif picker done for all users?  (You can always combine it with the sticker picker later)

IMO the gif picker is simply client-side, and the gif file should just be retrieved by the sender's client from the search service and sent as a file.  This will work on all clients, and avoids any worries about tracking bugs etc.

**Comment by beposec** [comment]
I second that. Just a client side gif picker would help a lot for now. Currently the iOS App is not even not able to play GIFs. They are converted to png... That's not what a messenger 2022 should do. 

For later it's sure good to implement a server side solution.

**Comment by ShadowJonathan** [comment]
The "GIFs are shown as PNGs" is an synapse as well as specification issue, see https://github.com/matrix-org/synapse/issues/1278 and https://github.com/matrix-org/matrix-spec/issues/453

**Comment by finncelta** [comment]
I 100% agree that this goes completely against what open source software should be. We have seen multiple PR about this exact issue and a lot of feedback on every request of GIF implementation that shows that users really want this feature.
This issue has also been 7 years in the making and all we have to show for it are designs "we don't want to double down on", a temporary client side solution like @iceHtwoO has built seems totally adequate for the moment and would make many people in the community happy and might even help garnering new users.

**Comment by J-tt** [comment]
As someone who has a bounty on getting this issue resolved, seeing the core team drop its priority makes me genuinely sad, especially when others have clearly put effort towards claiming the bounty.

Is there any way to put money towards an issue to get it actually prioritised?

**Comment by finncelta** [comment]
I've been talking with @iceHtwoO during the creation of his fork and to now see it just get dropped is kind of sad, we've been talking about maintaining a working fork that has the GIF integration, kind of like the SchildiChat guys do it. Other than that I don't see a way to get this into element without the core team implementing it themselves or becoming more open to the idea of client side GIFs.

**Comment by turt2live** [comment]
Hang on, there is definitely room for the community to contribute this. The current designs are far too complicated for someone to contribute though, which is why I'm sending it back to their queue.

**Comment by turt2live** [comment]
In an effort to add clarity: all the discussion about stickers and mixing the infrastructure is to convey that the current designs are not feasible. We may very well explore it in the future, but for right now it's obvious we need something easier to implement and support.

With PRs stacking up against the feature, we can't reasonably do code review though, so have asked design to take a look at the general problem rather than any individual PR. Once that's done, anyone is welcome to contribute it - the core team might look at it from a higher, more generic, point of view, however.

**Comment by moodler** [comment]
Is there anywhere we can see progress or timelines from the core team on this and perhaps contribute to UX testing?

I could send thousands of users your way if needed.

That said, I don't think it's a complex UI at all.  There are great examples to borrow from in WhatsApp, Twitter and Telegram.

**Comment by captainperth** [comment]
> Is there anywhere we can see progress or timelines from the core team on this and perhaps contribute to UX testing?
> 
> I could send thousands of users your way if needed.
> 
> That said, I don't think it's a complex UI at all. There are great examples to borrow from in WhatsApp, Twitter and Telegram.

Seconded, I work in a small team that uses element for all internal messaging. We'd love to contribute if it means helping getting this feature off the ground.

**Comment by moodler** [comment]
So, current status is, as I understand it, that the product team will be working on a recommended UI spec for this seven-year-old popular issue and publishing it here.

At that point, any one of us in the community who wants this done can implement it, and assuming the code is good, it will be integrated quickly.  OK, great.

Since this plan involves us (community) doing the bulk of the work and contributing resources, the respectful thing to do would be for someone from the core product team to publish some info on expected timelines.   We have zero visibility into their queue and ideas.  Is it a week?  A month?  A year?  Another 7 years?  ;-)   Just let us know please, thanks.  

Good communication makes good open source projects.

**Comment by beposec** [comment]
Bump! Any Update? @moodler asked important questions 😊

**Comment by iceHtwoO** [comment]
A easy way to send gifs may sound like a minor feature but it can make or break a chat client in my opinion. Please don't ghost us for another 7 years...

**Comment by J-tt** [comment]
Would especially like an update, since I put a cash bounty on the issue that is still just sitting in escrow :’(

**Comment by seanfarley** [comment]
> Since this plan involves us (community) doing the bulk of the work and contributing resources, the respectful thing to do would be for someone from the core product team to publish some info on expected timelines. We have zero visibility into their queue and ideas. Is it a week? A month? A year? Another 7 years? ;-) Just let us know please, thanks.

Looks like it'll be another 7 years 🙃

**Comment by turt2live** [comment]
I've raised this issue with the product team, again.

**Comment by seanfarley** [comment]
> I've raised this issue with the product team, again.

And with no visibility into that process, the community is left in the dark with no knowledge of the status nor a way to even contribute since all attempts have been closed.

**Comment by turt2live** [comment]
It's an internal process which attempts to get publicly-available feedback brought to the issue. While I appreciate this feature is desirable to a lot of people, please also be aware that the [team's roadmap is public](https://github.com/vector-im/roadmap/projects/1), and this issue is listed as "Future".

If it helps calm tensions a bit, I've asked the product team to explicitly raise their feedback here and for a timeline on when they think it could be implemented by. 

----

The true status of this is that community contributions won't make this go any faster, so from a code review perspective we're removing them from our backlog. The product and design teams need to engage with this issue to determine what the next steps are, be it community implementation or another team's implementation. 

Typically, a PR or random implementation will help the process, but in this case there's far too many considerations to make, which is why the development teams are rejecting PRs in favour of the product/design teams raising thoughts. Hopefully those teams will leave their feedback soon :)

**Comment by michelebugio** [comment]
I'm a bit concerned about what seems to be a bit of a disconnect between the team and users.

I understand there is a process, I understand the team is otherwise engaged.

However, without a gif keyboard, matrix cannot and will not be used by a plethora of people that, once nerds like us set them up (I set up a server just for a few friends, and many are doing that for small communities), would be very happy to use Element or any other interface, as they're very user friendly and they work well. But if there's no gif keyboard, they'd be in and out within a minute. If this is marked for "future", I can assume that there's either no interest for bigger adoption, or there's a disconnect, as this is a crucial feature for adoption, as silly as it may sound.

That being said, I do appreciate Travis for answering and raising the issue to the product team, thank you.

**Comment by turt2live** [comment]
fwiw, it's also important to distinguish between Element and Matrix here: while Element helps significantly to the uptake of Matrix, it is not the only one out there doing this :) 

On a similar note, on the Matrix side we've identified custom emoji/stickers as near-critically important, alongside threads and a few other features. The Matrix side will be pushing for this and it might have considerable impact on how a GIF keyboard gets implemented, for example.

**Comment by pszypowicz** [comment]
If using Giphy api is a security risk, this should really be ‚opt in’ feature then, with an option for advanced users to set up proxy. 

Do not let the ideal solution be an enemy of a good one.

**Comment by ItsEcholot** [comment]
> On a similar note, on the Matrix side we've identified custom emoji/stickers as near-critically important, alongside threads and a few other features. The Matrix side will be pushing for this and it might have considerable impact on how a GIF keyboard gets implemented, for example.

This is the same point that has been brought up in https://github.com/vector-im/element-meta/issues/321#issuecomment-1134918007 and the reason why it was reopened in the next comment: GIFs aren't stickers nor emojis. As mentioned in https://github.com/vector-im/element-meta/issues/321#issuecomment-1140423589 this could be a simple client only change which adds gif search

**Comment by seanfarley** [comment]
> Do not let the ideal solution be an enemy of a good one.

I was going to say this as well since we're almost at the *decade* mark. Trust has been utterly destroyed with this broken system of internal processes.

**Comment by turt2live** [comment]
> If using Giphy api is a security risk, this should really be ‚opt in’ feature then, with an option for advanced users to set up proxy.

Any API could be considered a privacy risk (note: not the same as a security risk, necessarily). Whether it's opt-in or out is largely a product decision, and the team will be asking that question: what is best for our users, and are there risks in taking that standpoint?

> Do not let the ideal solution be an enemy of a good one.

We categorically are not, however we have a different bar for what "good enough to ship" is.

> up in [#321 (comment)](https://github.com/vector-im/element-meta/issues/321#issuecomment-1134918007) and the reason why it was reopened in the next comment: GIFs aren't stickers nor emojis. As mentioned in [#321 (comment)](https://github.com/vector-im/element-meta/issues/321#issuecomment-1140423589) this could be a simple client only change which adds gif search

They indeed aren't stickers or emojis, but it's a highly related feature which might end up in the same place design-wise. What we don't want to do is design a feature only to redesign it completely a couple weeks later: we'd rather just get it mostly right the first time, given that sort of timeline.


> > Do not let the ideal solution be an enemy of a good one.
> 
> I was going to say this as well since we're almost at the _decade_ mark. Trust has been utterly destroyed with this broken system of internal processes.

Unfortunately, I don't think I can cover the 7 year path for how we go here or why it's better, but I do believe the changes made over the years (particularly the last 2 years) will have improved things to avoid this sort of thing happening again. The short history is that a GIF keyboard just hasn't beaten the value of the other work the available team bandwidth could do. We're in a much better position now to make it happen, though.

---

ftr: the product team acknowledged the internal ping to say they'll take a look. Nothing to report at the moment, but there is hope that we can have something happen in the short term to unblock this for the near term.

**Comment by pszypowicz** [comment]
> We categorically are not

![8DCCC2E1-A8D0-414A-88FE-C43F7F9B9956](https://user-images.githubusercontent.com/2733699/187993304-27c6bd1f-e273-4b45-a767-1bfc55558cc5.gif)

But on a serious note, I hope you are right :)

**Comment by ara4n** [comment]
The way to do this in a relatively privacy preserving manner would be to do something like https://signal.org/blog/giphy-experiment/

**Comment by scottwallacesh** [comment]
> The way to do this in a relatively privacy preserving manner would be to do something like https://signal.org/blog/giphy-experiment/

The downside to this, of course, is that we're relying on a centralised service to act as the "VPN" -- unless you're advocating that the Homeserver acts as this "VPN"?

**Comment by moodler** [comment]
While we're waiting for someone to unblock the development here, perhaps we can change this issue name?  "Add support for a GIF Keyboard" doesn't make a lot of sense and sounds niche, perhaps it could be "Add UI to find and send GIFs" or something.

(BTW my hacky awkward workaround for the moment is a private chatroom called "GIFs" where I've put all my commonly-used GIFs and I forward them into other rooms as needed)

**Comment by jakewb-b** [comment]
Hi folks - hopefully I can add some context from the product side, and offer to help a bit. 

We 100% see this an important and desirable feature. The difficulty is that it raises a bunch of questions that we'd have to answer if we were going to implement it ourselves: particularly around what privacy implications it has and how we manage those, and how customisable it needs to be. As Travis says, the feature is still fairly low down our internal list of priorities right now so no one's had capacity to take time away from other things and figure out all those answers. 

At the same time, we've been reluctant to merge PRs that deliver this feature without us having answered those questions, as we're ultimately responsible for the feature once it's merged. I appreciate that's causing frustration, as people who want gifs are stuck between a rock and a hard place. 

Anyway, I've done some reading through the discussion here and on the PRs that have previously been submitted, and I think if I put all the pieces together and get some agreement internally, we probably do have most of the answers we need or can quickly arrive at them. So, if you could bear with me for a week or so, I can come back with some guidelines for an implementation that we would accept, and I'll commit some time to working with anyone who implements it to ensure that it's not blocked on product review once it comes time to submit the PR.

**Comment by arkitoure** [comment]
Thank you @jakewb-b . Ive been lurking on this thread for some time hoping for some forward movement, pardon me for swooping in like so but our user base does desire to have this feature at the very least of all image based add-ons. Stickers quite frankly no one cares about and polling the community it seems like GIFs are higher up on the list than anything else.

We moved our community from discord to matrix a while ago and would love to offer some of the more lively features for a bit more candid energy in our conversations. I believe the desire comes simply from the need to express one self in the most diverse way and GIFs grant a wide array of expression beyond words.

We do however appreciate the security focused angle here as minimal bells and whistles provides. Looking forward to what comes next and contributing!

**Comment by pagdot** [comment]
> Stickers quite frankly no one cares about and polling the community it seems like GIFs are higher up on the list than anything else

Wouldn't say that. Stickers are a great way to add customized reactions similar to emojis. Although I don't use them with a big reason being that they aren't supported by element. I've seen them used for reaction memes in a small community adding a bit more fun when you not just post a generic face-palm emoji or even gif, but a sticker with one member of the community face palming.

I'm also the opinion that gifs should have a higher priority, but also see the potential in stickers :)

**Comment by arkitoure** [comment]
@pagdot, yes I would agree with you. While this is something I'd also like to see developed to the level of Telegram, for instance, in our community people are not so keen on them compared to GIFs. But by all means keep both features on the table for short to mid term implementation, GIF's being 1st up.

**Comment by jakewb-b** [comment]
Afternoon everyone. Thanks for your patience with this. I've been giving this some more thought, and I'll try to lay out the key points we'd be concerned about from a product perspective, to hopefully give some guidance to people who want to implement this and make it more likely that we could merge a community-contributed PR in future.

First off - @gaelledel provided some designs including behaviour specs here: https://www.figma.com/file/ifIvZA4XIP1JTyndHFr5Ub/Community-PRs?node-id=10%3A7689
These remain a good guide to what we would like to see from a UI perspective. If a contribution manages to stick quite closely to these designs, it should reduce the likelihood of a future PR getting blocked on design review. 

From a product perspective, here are some of the previously unanswered questions that I've tried to answer:

- Should any given service (e.g. Giphy) be the default / the only option? 
   - We're comfortable with the provider being fixed if this feature releases in labs. For it to be de-labsed, we'd want to see a way that a user can, reasonably easily and intuitively, select an alternate service. It would be ok to have a default choice, though. 
- Are we concerned about leaking data to the gif provider?
   - Preventing leaking IP addresses to the provider would be a hard requirement for the feature. We don't have the tech bandwidth right now to investigate the best way to do this, but we're happy to jump in and assist/comment if people have proposals. 
- Do we need settings to disable the feature, disable viewing gifs, disallow sending gifs in a room, etc?
   - The designs and behaviour spec linked above include a user preference to enable the gif picker, and a new room setting to control who can send gifs in the room. However, these designs were done before we settled on the point above about leaking IP addresses so those settings _may_ not be required, and as a rule I'd strongly prefer to avoid proliferation of very specific settings. So, it's probably better if I express the problems we wanted to solve, and open it to you to see if there are ways to do this that avoid or reduce the use of extra settings.
   - 1. We want to both prevent users leaking their data to a third party, and reassure them that we're preventing this - or at least that they have control over it.
   - 2. We would prefer to avoid users being obliged to download received gifs if they don't want to, especially if they're on limited connections.
   - 3. We want to help room admins to set the 'tone' they want for their room, and they may mean avoiding the timeline becoming full of gifs. 

I hope this is some help. I can't promise I've thought of every single detail or nuance here, so I'd encourage people to ask questions and seek clarification if needed before submitting PRs, so that hopefully we can successfully get something merged.

**Comment by seanfarley** [comment]
> Do we need settings to disable the feature, disable viewing gifs, disallow sending gifs in a room, etc?

This seems orthogonal to the issue at hand. Currently, I can just drag-and-drop a gif and the user will receive it no matter what (submit `b`). I understand how something like this could slip into this issue but I strongly suggest making it its own, separate issue.

**Comment by pagdot** [comment]
I could imagine designing a gif api and implementing a proxy for giphy or other similar services with this API. Additionally, a privacy oriented selfhostable gif service could be implemented.

Giphy or other providers would only know the ip from the proxy and the proxy could be implemented privacy friendly. Other chat apps could also implement this api client side to support gif proxies.

A simple proxy sever would probably be pretty straightforward to implement because It would only need to forward the requests with a search string and return a stream of images. Apart from concurrency/connection limits, protection from attacks could be implemented at a reverse proxy level.

**Comment by beposec** [comment]
> This really feels like something that should be in Element, I've put a bounty on it:
> 
> https://www.bountysource.com/issues/28576328-we-need-a-gif-keyboard-toooooo-11-1

Increased the bounty :-)

**Comment by anoadragon453** [comment]
Signal has a server-side proxied giphy search feature:

* https://signal.org/blog/giphy-experiment/
* https://signal.org/blog/signal-and-giphy-update/ 

I looked into how this works and whether there was any code we could re-use. Instead of a gif-searching-specific proxy, Signal look to have implemented a generic proxy for any request from the client: https://github.com/signalapp/Signal-TLS-Proxy. I think this is then used to proxy requests to giphy from the client.

https://github.com/signalapp/Signal-Android/blob/c4bc2162f23e0fd6bc25941af8fb7454d91a4a35/app/src/main/java/org/thoughtcrime/securesms/net/ContentProxySelector.java#L25
https://github.com/signalapp/Signal-Android/blob/c4bc2162f23e0fd6bc25941af8fb7454d91a4a35/app/src/main/java/org/thoughtcrime/securesms/giph/mp4/GiphyMp4PagedDataSource.java#L32-L35

This approach works and is useful for generally proxying requests from the client... however I'm unsure what other features would actually require such proxied requests. And this does require the client to specifically implement giphy-related code, rather than some generic "external media search" API (I could imagine modules on the backend for giphy, imgur, etc).

**Comment by hex-m** [comment]
> however I'm unsure what other features would actually require such proxied requests

Link previews come to mind. This was already discussed here: https://github.com/vector-im/element-meta/issues/1138

Maybe Widgets too? I assume the server hosting a widget learns the IPs of all the clients in the room in the current setup.

And the Map-Tile Server. Most users probably use the default one. That service could learn who is in a room together because they viewed the same coordinates. I already suggested using their proxy solution for maps in the [Signal Community](https://community.signalusers.org/t/use-openstreetmap-instead-of-google-maps/5434/20).

**Comment by pszypowicz** [comment]
Something like Signal TLS Proxy seems to be server based. Should we have an MSC first before we jump into implementation?

Regarding implementation:
I think most of the people deploy synapse behind a LB, what if we configure additional path where we reuse Signal TLS Proxy.

Client side GIF picker, could as a component, discover with some sort of a health check if that "feature" (Proxy) is available on the home server and only then present itself?

**Comment by moodler** [comment]
Pardon me for charging into this rabbithole :) but I have a question about this assumption made above "Preventing leaking IP addresses to the provider would be a hard requirement for the feature."  

If I'm using a client-side implementation (where the gif file is sent as a file instead of a link), what actually is the perceived problem with giphy knowing my IP address when I search their site for a GIF?   I don't see this explained anywhere above.

AFAICT, it's exactly the same as if I went to https://giphy.com (or any site) on a web browser.

**Comment by turt2live** [comment]
Our typical userbase would be extremely surprised if their IP was sent to Giphy directly, even if we advertised the feature as powered by Giphy.

**Comment by moodler** [comment]
I doubt that at all.  

It's exactly like using the web.  People who REALLY worry about would be using VPNs, I assume.  Or just not sending GIFs.

What exactly do you think Giphy could possibly do with the IP of your phone/tablet/laptop?

**Comment by turt2live** [comment]
It's not what they *could* do, it's the perceived risk and reputational damage of the project. We're not comfortable with leaking people's IP to Giphy, and our users generally expect us to hold that opinion (rightly so, imo).

**Comment by jakewb-b** [comment]
> People who REALLY worry about would be using VPNs

People who *really* worry about it might, we hope, be using Element. Our users expect us to provide a level of privacy above that offered by regular use of the web.

**Comment by moodler** [comment]
If you want to build complex proxy systems around simple GIFs, of course you can, but about "levels of privacy", I will just point out that the makers of the many clients recommended on https://matrix.org/clients/ (and the ones not linked) also have the ability to see and collect client IPs, and in many cases also to scrape and keep your username and password

**Comment by ShadowJonathan** [comment]
The question of privacy isn't about second parties (clients) gaining access to leak it, it's about the third parties (giphy, in this case).

A user must always implicitly trust a second party for this information, but it's not necessary for a third party to do so, especially for a trivial reason such as this.

But this is _especially_ important when the company currently itching to buy giphy is one of the most notorious privacy violators in the world, right now. So in addition to user expectation, this is not something that should be thrown in without caution.

**Comment by beposec** [comment]
The bounty is now at $272.00USD
https://app.bountysource.com/issues/28576328-we-need-a-gif-keyboard-toooooo-11-1

**Comment by cesar-cardinale** [comment]
### FYI
I just create a widget to let you search a GIF in GIPHY and send it to your room when clicking on it, you just need to add it with :
`/addwidget https://matrix.cesarcardinale.fr/#/?widgetId=$matrix_widget_id&userId=$matrix_user_id`
It only work on desktop and web, the https://github.com/matrix-org/matrix-widget-api doesn't work on mobile.
And you must enable the URL preview because Matrix only shows images that are hosted on the server, and I don't want to host all the GIF of GIPHY !

Demo :

https://user-images.githubusercontent.com/37698963/212481056-e3d26e56-46fb-4c11-88bc-89ce60634994.mp4

**Comment by arkitoure** [comment]
Nice work @cesar-cardinale, good to see some progress on this. To note, a mobile solution shouldn't be needed anyway as most, if not all, mobile touch keyboards will have GIFs integrated. So desktop and web is just what is needed!

**Comment by Melechtna** [comment]
I'd honestly rather it actually post the gif than do a link, but it's a start. :)

**Comment by cesar-cardinale** [comment]
@Melechtna me too, but Element only shows images that are hosted on Matrix with an url like this : ´mxc:...´

**Comment by Melechtna** [comment]
> @Melechtna me too, but Element only shows images that are hosted on Matrix with an url like this : ´mxc:...´

Don't supposed there's anyway to throw it into a wrapper and make it think it's coming from that source?

Might be outside of the scope, being something the devs just need to properly implement if your method is accepted.

**Comment by cesar-cardinale** [comment]
> Don't supposed there's anyway to throw it into a wrapper and make it think it's coming from that source?
> 
> 
> 
> Might be outside of the scope, being something the devs just need to properly implement if your method is accepted.

Maybe but this issue is opened since 7 years so I don't think that a dev is going to do something for it.

Maybe if I have time, in the future, I will try to make a PR, but in term of security it could be a problem, like you can wrap a fake image or other things and run a script on each Element devices that show this "image".

**Comment by seanfarley** [comment]
> Maybe if I have time, in the future, I will try to make a PR, but in term of security it could be a problem, like you can wrap a fake image or other things and run a script on each Element devices that show this "image".

It wouldn't be accepted because the current status of this issue, as far as I understand it, is stuck in a "waiting for our designers" but the last designer to chime in did so [four months ago](https://github.com/vector-im/element-meta/issues/321#issuecomment-1243775736) and only rehashed points made (many times, in fact) in this thread previously. The Figma board that is linked is private so I don't know how a potential contributor is even supposed to follow it.

The Matrix team, once again, shows that they lack the leadership and resources to manage such a large projects so I suggest just closing this issue as `WONT FIX`. Potential teams looking to adopt matrix should take heed as to how these embarrassingly long-standing issues are never properly addressed.

**Comment by turt2live** [comment]
To be clear, this is with the Element team rather than the Matrix team. It's also really not a lack of leadership or resources, but rather a deliberate decision to prioritize everything else above this feature. This is obviously not agreeable to everyone.

It's great to see that others are taking a look at trying to implement this, though as mentioned previously: it's difficult to justify code review while there's security and privacy concerns on the table. Discussion towards solving those problems is more appreciated because we might be able to work around the design team having other priorities once they're sufficiently solved. Unfortunately, a widget has the same sort of risks we're worried about regarding user privacy, in addition to being an unreliable technical foundation for such a feature (the widget API is highly unstable).

**Comment by seanfarley** [comment]
And again, to be clear, the <whatever team> has had 7+ years to tighten the scope on this issue to prevent it from becoming bloated (stop trying to wedge this in with the widget API) along with plenty of discussions in this thread to solve the security and privacy issues. It's a simple issue that, quite literally, every other chat client has solved and shows the utter incompetence of this team.

For others looking for alternatives to matrix, ejabberd is a solid project that has better leadership along with [adding support for matrix recently](https://www.process-one.net/blog/matrix-protocol-added-to-ejabberd/). Our teams are already in the process of switching and highly recommend it.

**Comment by turt2live** [comment]
I haven't seen solutions to the problems, though there is discussion. If ~~I'm blind~~ I've missed something, we're more than happy to see a PR for this. 

> stop trying to wedge this in with the widget API

I believe I said this *shouldn't* use the widget API, unless you mean to say we should be doing so? (in which case, PRs on the mobile clients to support the features of the widget API needed would be great).

**Comment by seanfarley** [comment]
> If I'm blind

Let's try not to be ableist, please.

**Comment by turt2live** [comment]
My apologies, though I would appreciate an answer to my question.

**Comment by arkitoure** [comment]
I seriously respect the conservative approach to privacy & security. I just wish this one thing would get hashed out officially so our communities can have at least one of the extra features they want. I didn't even realise this issue has been open since 2015. Element is such a nice project, I believe the lot of us want the best for it. Any bounty on this issue?

**Comment by turt2live** [comment]
The designs appear to be based on an older template of the app, but here's the key parts from the Figma:

![image](https://user-images.githubusercontent.com/1190097/213309025-3186efa6-b0b6-4b59-bb84-c5796e323643.png)
![image](https://user-images.githubusercontent.com/1190097/213309050-9ac0e1dc-be34-4738-96d4-63dbb11aee61.png)
![image](https://user-images.githubusercontent.com/1190097/213309123-e64fe5c0-6ba0-4ba5-9516-62edfb17086c.png)
![image](https://user-images.githubusercontent.com/1190097/213309159-20560121-d9b5-4535-b9d3-4990e8c632ab.png)

The last might not be possible at this stage, so would encourage not worrying about it too much.

**Comment by moodler** [comment]
I still think the best way to solve this is to:

1. Use [tenor.com](https://tenor.com) by default instead of giphy, since giphy's ownership has issues (ideally a user could choose their favourite GIF source)
2. As per the interface above, make it a USER setting to allow sending of gifs, off by default.  Turning it on includes a small warning that tenor.com will be able to see your search strings and IP.
3. Selecting a gif (and again, the mockup above is great) will post the gif FILE to the chat (not a link).  This way it works on all clients right away without any information being leaked.

**Comment by MightyCreak** [comment]
@turt2live Not sure the room settings for GIF integration is relevant (in the last mockup you shared), because I can always save a gif on disk and send it as an attachment in Element... It will be displayed as a gif image that can be played, although it's not coming from Tenor or Giphy. So this would be allowed, but using the integrated button in the composer wouldn't? Seems like a weird design, wouldn't it?

I think moderators would find it interesting to filter out GIF files (no matter their source) in some rooms, but that would be a bit more difficult to implement when users will be able to send rich messages (mixing text and media, as we can also see in the first mockup you shared 👀😉)

**Comment by turt2live** [comment]
@moodler I don't think we're tied to a provider, so go forth and implement it agnostic.

@moodler @MightyCreak the room permissions setting is more tied to Extensible Events than GIFs - the design is considered "exploratory" and not final, which largely means it hasn't had full technical review yet. For the scope of this issue, that permission can be ignored.

**Comment by cesar-cardinale** [comment]
> 3. Selecting a gif (and again, the mockup above is great) will post the gif FILE to the chat (not a link).  This way it works on all clients right away without any information being leaked.

So you are saying that if a lot of people send GIF, you will save them on your server? What a waste of disk space ...

**Comment by t3chguy** [comment]
@cesar-cardinale the alternative is for your client to leak your IP address to the GIF hoster.

**Comment by cesar-cardinale** [comment]
The best way to do that would be to leak the IP address of the server and not of the client ...

**Comment by t3chguy** [comment]
@cesar-cardinale yup, and the server would cache, which still wastes disk space. Synapse caches indefinitely.

**Comment by cesar-cardinale** [comment]
Yeah, but if you make backup of your server, there will not those GIF in the backup, only the links, I think it would be a better way to implement that. But yes the cache would take more space than just a link but less than saving the GIF on the server.

**Comment by turt2live** [comment]
For clarity: we're not changing how media (GIFs) are distributed in Matrix. That's something for a completely different issue.

Please keep the discussion around the GIF keyboard specifically. We've already discussed the implications of where your IP goes in this issue.

**Comment by t3chguy** [comment]
@cesar-cardinale looks like you've never backed up a Synapse. URL and remote media caches are definitely right next to your local media files. If you simply don't include them in the backup they will never be refetched until you manually fix your database to purge the cache.

**Comment by Samgarr** [comment]
> Any bounty on this issue?


Bounty is now $552 USD

https://app.bountysource.com/issues/28576328-we-need-a-gif-keyboard-toooooo-11-1

**Comment by weblionCH** [comment]
This topic is now unresolved since nearly 8 years, when I see the starting date of 2015 correctly. Is there any realistic chance, that this finds its way into a release for all the clients this year? From a communication standpoint it's (a bit) a shame to not provide a proper solution for stickers & GIFs. Bounty is already over $600… please!🥲

**Comment by dillfrescott** [comment]
$652 lol. Hey chatgpt make me a gif picker for element! 🤣

**Comment by moodler** [comment]
Workaround, if you are on MacOS, you can use [GIF Keyboard for Mac](https://tenor.com/mac) 

It gives you a desktop menu where you can search and copy a gif, and then you can just paste it into Element (or anything else).

I still think it's lame for a messenger in 2023 not to have native support for these things, but that's where we are.

![tenor-136213481](https://github.com/vector-im/element-meta/assets/88039935/752b2876-5a2e-447d-a9ed-717bc99d0b25)

**Comment by VPaulV** [comment]
Well with `url_preview_enabled: true` in synapse we can use the widget provided by @cesar-cardinale. 
Funny that this is still an issue for a messenger client in 2023 xD
![image](https://github.com/vector-im/element-meta/assets/5863225/85f6757d-2731-44f6-9622-43da2117a714)

Note, you should also enable url previews in room settings

**Comment by ranenvious** [comment]
This feature is honestly the main reason me and my friend group are still on discord. It seems minor, but sarcastic gif responses are genuinely a large part of our little friend group culture. It may not technically be a huge feature in terms of what it offers, but having the ability to search for, save, and send gifs all super easily is the catalyst for so many jokes that just flat out aren't possible to deliver right through text alone.

**Comment by williamkray** [comment]
i know this is definitely not the same, but there are bots that will fetch a random gif for you... shameless plug for my [maubot plugin](https://github.com/williamkray/maubot-gifme) which allows you to set either giphy or tenor as the fallback source, and whenever it returns something good (i've had better luck with tenor) you can save it to your own personal database. this means no managing a collection of gifs manually, but still getting the things you are expecting when you expect them.

far from any sort of search interface but honestly half the fun is seeing if the bot will pick something good or not :laughing:

**Comment by Lamborge** [comment]
> The designs appear to be based on an older template of the app, but here's the key parts from the Figma:
> 
> ![image](https://user-images.githubusercontent.com/1190097/213309025-3186efa6-b0b6-4b59-bb84-c5796e323643.png) ![image](https://user-images.githubusercontent.com/1190097/213309050-9ac0e1dc-be34-4738-96d4-63dbb11aee61.png) ![image](https://user-images.githubusercontent.com/1190097/213309123-e64fe5c0-6ba0-4ba5-9516-62edfb17086c.png) ![image](https://user-images.githubusercontent.com/1190097/213309159-20560121-d9b5-4535-b9d3-4990e8c632ab.png)
> 
> The last might not be possible at this stage, so would encourage not worrying about it too much.

can you tell me how soon this will be added to the element? Or is there some experimental version of the element with this functionality?

**Comment by J-tt** [comment]
@Lamborge I wouldn't hold my breath, beside the Figma designs there has been little interest from the core team in ever making this a reality.

**Comment by iceHtwoO** [comment]
@Lamborge Multiple people have implemented this feature and all PRs where closed. 
https://github.com/matrix-org/matrix-react-sdk/pull/8712
https://github.com/matrix-org/matrix-react-sdk/pull/5814

**Comment by arkitoure** [comment]
@iceHtwoO the web app or os clients?

**Comment by iceHtwoO** [comment]
@arkitoure its some time since I've touched this project but i believe it can be used for both.

**Comment by Lamborge** [comment]
> @Lamborge Multiple people have implemented this feature and all PRs where closed. 

Why were they closed? These PRs look very impressive!

**Comment by iceHtwoO** [comment]
> > @Lamborge Multiple people have implemented this feature and all PRs where closed.
> 
> Why were they closed? These PRs look very impressive!

"The feature requires design and product review and is generally quite large in size - we might try to take the work on ourselves if/when product/design get a chance to take a look." - [turt2live](https://github.com/turt2live)

**Comment by Lamborge** [comment]
> > > @Lamborge Multiple people have implemented this feature and all PRs where closed.
> > 
> > 
> > Why were they closed? These PRs look very impressive!
> 
> "The feature requires design and product review and is generally quite large in size - we might try to take the work on ourselves if/when product/design get a chance to take a look." - [turt2live](https://github.com/turt2live)

meh

**Comment by michelebugio** [comment]
each time I get a notification on this topic, I laugh. This is one of the best jokes I've ever seen at this point. 
I wonder if an ambitious and delicate feature such as this will be ever worked out. We must approach the best minds in the world in order to be able to add sarcastic animated faces and funny cats from the privacy of our basements.
Truly one of the biggest challenges for the human race to accomplish.

**Comment by VPaulV** [comment]
@Jayddo there is a better one, how about 6 years we cannot get a custom notification sound (or at least a better one), despite all the PRs?

**Comment by JiffyCat** [comment]
At this point, this is to be expected from the Element team. Another highly requested feature ([custom emotes](https://github.com/vector-im/element-meta/issues/339)) is blocked even with multiple PRs.

**Comment by ghost** [comment]
It's a real shame that all "fun" features get no attention and are even actively rejected by the developers.
You can't have widespread adoption when you're lacking behind on features that make other chat platforms fun to use.
I couldn't convince my friend group to move to matrix because of the lack of gifs and custom emojis, and honestly, they're right. As much as i want to move away from Discord, they at least have good features.

**Comment by nebulade** [comment]
Really looking forward to this as well.

**Comment by TouchstoneTheDev** [comment]
> This topic is now unresolved since nearly 8 years, when I see the starting date of 2015 correctly. Is there any realistic chance, that this finds its way into a release for all the clients this year? From a communication standpoint it's (a bit) a shame to not provide a proper solution for stickers & GIFs. Bounty is already over $600… please!🥲

Duhh bountysource is goneeee

**Comment by J-tt** [comment]
Oh yikes 

https://github.com/bountysource/core/issues/1586

**Comment by TouchstoneTheDev** [comment]
> Oh yikes
> 
> [bountysource/core#1586](https://github.com/bountysource/core/issues/1586)

Others are wasted for life

**Comment by crabdancing** [comment]
> It's a real shame that all "fun" features get no attention and are even actively rejected by the developers. You can't have widespread adoption when you're lacking behind on features that make other chat platforms fun to use. I couldn't convince my friend group to move to matrix because of the lack of gifs and custom emojis, and honestly, they're right. As much as i want to move away from Discord, they at least have good features.

I completely agree with this. Like it or not, whether you think it's frivolous or not, people are not generally driven by what's better under the hood. They're driven by what lets them communicate as viscerally and comfortably as possible. Trying to get a product like Element/Matrix truly mass-adopted while ignoring features like this, is like insisting on making a trench flowing _uphill_, and hoping the water will cooperate with you. I want to see this tech stack _replace_ Discord. It is technically better in every way that matters, it is less predatory, more respectful of privacy and freedom rights, and has a business model that seems entirely sustainable. But we need to adopt some level of 'frivolity' to make that a reality.

Edit: typo

**Comment by weblionCH** [comment]
> > It's a real shame that all "fun" features get no attention and are even actively rejected by the developers. You can't have widespread adoption when you're lacking behind on features that make other chat platforms fun to use. I couldn't convince my friend group to move to matrix because of the lack of gifs and custom emojis, and honestly, they're right. As much as i want to move away from Discord, they at least have good features.
> 
> I completely agree with this. Like it or not, whether you think it's frivolous or not, people are not generally driven by what's better under the hood. They're driven by what lets them communicate as viscerally and comfortably as possible. Trying to get a product like Element/Matrix truly mass-adopted while ignoring features like this, is like insisting on making a trench flowing _uphill_, and hoping the water will cooperate with you. I want to see this tech stack _replace_ Matrix. It is technically better in every way that matters, it is less predatory, more respectful of privacy and freedom rights, and has a business model that seems entirely sustainable. But we need to adopt some level of 'frivolity' to make that a reality.

I couldn't agree more! This needs to be implemented to stay competitive on the market. Period.

**Comment by dillfrescott** [comment]
You guys are asking element hq to add a useful feature to their clients. Isnt that against the ToS?

**Comment by MightyCreak** [comment]
While I would like to have this feature too, I really don't like the recent comments I'm seeing here. I find them disrespectful.

I am fairly sure that Element developers know and understand the relevancy of such feature, but they also have other important features to implement (say [redo all the mobile apps](https://element.io/labs/element-x), for instance...). Also, they have clients (those that are actually paying them and allows them to [barely stay afloat](https://element.io/blog/synapse-now-lives-at-github-com-element-hq-synapse/)) for specific features and these features are, apparently, not about being able to add GIFs in their messages.

If you really can't bear the fact that element web/desktop don't provide you with a quick way to add GIFs in your texts, maybe try [other Matrix clients](https://matrix.org/ecosystem/clients/)?

**Comment by spaetz** [comment]
Not wanting to increase the noise ratio, but this is not really an element-meta issue, as mobile keyboards on e.g. Android feature their own GIF picker, which work just fine with element mobile clients. (I verified), so this is mostly limited to element-web/desktop.

**Comment by iceHtwoO** [comment]
> Not wanting to increase the noise ratio, but this is not really an element-meta issue, as mobile keyboards on e.g. Android feature their own GIF picker, which work just fine with element mobile clients. (I verified), so this is mostly limited to element-web/desktop.

iOS doesn't have this feature by default

**Comment by r1lax** [comment]
We are looking forward to it as well here :)

**Comment by Mesmaroth** [comment]
Any updates on this? It's been 9 years since this issue was started.

**Comment by ara4n** [comment]
It's stuck in the same state as https://github.com/element-hq/element-meta/issues/339#issuecomment-2052188082.

**Comment by k3dom** [comment]
Seeing how old this issue it almost seems ridiculous even making this comment but not having this feature is honestly a big bummer. There is simply no way matrix will get wide spread adoption if its most popular client is missing even such simple features that, at this point, are well established in online chatting culture. Would be awesome seeing this added!

**Comment by TheBinaryShadow** [comment]
> Seeing how old this issue it almost seems ridiculous even making this comment but not having this feature is honestly a big bummer. There is simply no way matrix will get wide spread adoption if its most popular client is missing even such simple features that, at this point, are well established in online chatting culture. Would be awesome seeing this added!

Amen. How are we expected to migrate and entrust any kind of workflow or mission critical task to a client that can't get GIF search in 2024. 

Yes, I am serious. 

Stop trying to build the roof of this house, there ain't no base to hold it.

**Comment by k3dom** [comment]
Yeah I mean this sadly just seems to be a good example of an open source business strategy gone wrong (at least for anybody that's not NATO or the German government). You are gonna be hard-pressed to find an open source project of this size where the requirement to merge a PR is ["multiple hundred thousand $"](https://github.com/element-hq/element-meta/issues/339#issuecomment-2057475632) for a feature that not only has been requested 9 years ago but is also dead simple and much wanted by a large amount of the user base.

That being said, complaining in GitHub issues probably isn't gonna do anything good either. Just sad to see really as this is honestly an awesome project otherwise.

**Comment by michelebugio** [comment]
While understanding the priorities and constraints is important, this kind of feature should be almost an after-thought and at the same time it is one of the main requirements for mass adoption, so seeing as how it is gated under paid requests, and even then, less prioritized than governments requests, while reading how matrix still strives for mass adoption, makes it obvious how there's a big disconnect between what the software wants to do and what it's working towards.

Mass adoption means catering towards the end users' needs that change with time, and gif search has been a staple since... 2020? something around that, so it'd be expected within.. one year of it becoming a standard, let's say. There will be more things in the future, for sure, and being on point is crucial to be adopted and stay adopted. This shows that there's no ball being dropped, because there's no ball.

For us simple users, it simply means matrix cannot be used and we need to look elsewhere for our communities, small businesses and what not. Mass adoption as a result will not be there and this will be a great software used in specific environments. Nothing wrong with that.

It's just a bit sad that a simple thing such as gif search becomes the divider between mass adoption and niche use, but that's where we're at.

This all being said, it does not mean to discourage the effort done, and while I may disagree with how this feature is being actively ignored, I do appreciate the kind of work that I would not be able to do, so cheers for that, Matrix and Elements are great pieces of software.

**Comment by arkitoure** [comment]
My issue is that Matrix/Element is spidered into our whole community setup and is difficult to switch now. People are constantly going on about DIscordish features but we wanted to use a decentralised solution and Matrix at the time was the best fit. I still think the tech is highly competitive but this more necessity based approach does split opinions.

With security and decentralisation at the forefront, some argue that these bells and whistles are childish and simply not needed. While others, the more Discord type folk, joke about how they cannot use custom emoj and gifs. Personally Im kind of in the middle of those lines of thinking due to having bigger fish to fry in time:production ratio. I do lean more on the security/decentralisation side however.

Developing good software is hard work and costs money and I know this is a big factor but I still have hope that the project will elevate beyond. I hope to also contribute more here once I get through the stack of todo's on my desk.

**Comment by hegdenischay** [comment]
Just wanted to chime in and say that since there's possibly a privacy problem with integrating a GIF picker straight into element, I've created a [PR to maunium/stickerpicker](https://github.com/maunium/stickerpicker/pull/75) to add it there, so whoever wants this feature could use it through the existing stickerpicker widget. It's also disabled by default, and requires a Giphy API key to be used. (You can get it by snooping around network requests on the giphy website.)

It's definitely slower than what you would expect, but I think that's probably because I'm bad at javascript rather than anything inherently wrong in the concept.

It currently uses Giphy, maybe someone can contribute Tenor search as well.

**Comment by sascha-wi** [comment]
Can somebody give an clear overview of the current status of trying to implement a gif provider/integration into matrix / element?

1. Is it possible?
2. Through which means?
3. What are other options?

**Comment by hegdenischay** [comment]
@sascha-wi If you want a GIF picker in element, use [maunium/stickerpicker](https://github.com/maunium/stickerpicker/) as a widget. This uses giphy, and it should be pretty quick now. 

If you want GIF search in a matrix client, [commet.chat](https://commet.chat/) implements it with Tenor (IIRC). Commet.chat is pretty new, however, and I have not personally used it, so I cannot vouch for how well their GIF implementation works.

**Comment by spaetz** [comment]
There is also gifme, the maubot which is able to search and display giphy and tenor pics. (At least) Fluffychat supports custom emojis and many Android's keyboard contains sticker functionality...

**Comment by sascha-wi** [comment]
> @sascha-wi If you want a GIF picker in element, use [maunium/stickerpicker](https://github.com/maunium/stickerpicker/) as a widget. This uses giphy, and it should be pretty quick now.
> 
> If you want GIF search in a matrix client, [commet.chat](https://commet.chat/) implements it with Tenor (IIRC). Commet.chat is pretty new, however, and I have not personally used it, so I cannot vouch for how well their GIF implementation works.

Does maunium also work for the element desktop app? I have to create those "sticker packs" myself according to the documentation/installation video from Brodie Robertson. Thats not what I am looking for, I am trying to integrate a gif picker for all of the users of my matrix server.

**Comment by hegdenischay** [comment]
> > @sascha-wi If you want a GIF picker in element, use [maunium/stickerpicker](https://github.com/maunium/stickerpicker/) as a widget. This uses giphy, and it should be pretty quick now.
> > If you want GIF search in a matrix client, [commet.chat](https://commet.chat/) implements it with Tenor (IIRC). Commet.chat is pretty new, however, and I have not personally used it, so I cannot vouch for how well their GIF implementation works.
> 
> Does maunium also work for the element desktop app? I have to create those "sticker packs" myself according to the documentation/installation video from Brodie Robertson. Thats not what I am looking for, I am trying to integrate a gif picker for all of the users of my matrix server.

the GIF picker is enabled by default now, and you should simply be able to use Giphy GIFs without having to add any new stickers. If your users want custom stickers later, you could add them to one instance and every user would have that.

**Comment by sascha-wi** [comment]
> > > @sascha-wi If you want a GIF picker in element, use [maunium/stickerpicker](https://github.com/maunium/stickerpicker/) as a widget. This uses giphy, and it should be pretty quick now.
> > > If you want GIF search in a matrix client, [commet.chat](https://commet.chat/) implements it with Tenor (IIRC). Commet.chat is pretty new, however, and I have not personally used it, so I cannot vouch for how well their GIF implementation works.
> > 
> > 
> > Does maunium also work for the element desktop app? I have to create those "sticker packs" myself according to the documentation/installation video from Brodie Robertson. Thats not what I am looking for, I am trying to integrate a gif picker for all of the users of my matrix server.
> 
> the GIF picker is enabled by default now, and you should simply be able to use Giphy GIFs without having to add any new stickers. If your users want custom stickers later, you could add them to one instance and every user would have that.

This was referring to [stickerpicker](https://github.com/maunium/stickerpicker)?

**Comment by hegdenischay** [comment]
> > > > @sascha-wi If you want a GIF picker in element, use [maunium/stickerpicker](https://github.com/maunium/stickerpicker/) as a widget. This uses giphy, and it should be pretty quick now.
> > > > If you want GIF search in a matrix client, [commet.chat](https://commet.chat/) implements it with Tenor (IIRC). Commet.chat is pretty new, however, and I have not personally used it, so I cannot vouch for how well their GIF implementation works.
> > > 
> > > 
> > > Does maunium also work for the element desktop app? I have to create those "sticker packs" myself according to the documentation/installation video from Brodie Robertson. Thats not what I am looking for, I am trying to integrate a gif picker for all of the users of my matrix server.
> > 
> > 
> > the GIF picker is enabled by default now, and you should simply be able to use Giphy GIFs without having to add any new stickers. If your users want custom stickers later, you could add them to one instance and every user would have that.
> 
> This was referring to [stickerpicker](https://github.com/maunium/stickerpicker)?

Yes.

**Comment by zerocode97** [comment]
Any updates to this?

**Comment by Destinyg133** [comment]
Still nothing?

**Comment by IAMBUDE** [comment]
After 9 years and 9 months, how is this still not a feature?

**Comment by nebulade** [comment]
https://github.com/matrix-org/matrix-spec-proposals/pull/4230 to better handle animated images / gifs with clients overall will certainly help to go in that direction. It much improved the gif handling in [fractal](https://gitlab.gnome.org/World/fractal)

**Comment by moodler** [comment]
While nice, I don't see how the flag would add much - gifs that I manually paste/upload into Element desktop already seem to propagate and appear fine for everyone.    This issue is more about the user interface for selecting them within all the Element apps so that ordinary non-nerds can enjoy a full chat experience. 

However, I'm resigned to the fact that Element dev team as a whole is consumed with different things.

I can't help but wonder if all the community energy spread into making umpteen different incomplete clients for Matrix was more focussed on just one or two fantastic clients then how much better and more competitive Matrix would be as an ecosystem.

**Comment by Borrower543** [comment]
>gifs that I manually paste/upload into Element desktop already seem to propagate and appear fine for everyone. This issue is more about the user interface for selecting them within all the Element apps so that ordinary non-nerds can enjoy a full chat experience.

Minor correction, gifs pasted into Element desktop don't animate. You need to manually save to device and manually upload. (Gifs entered through gboard/etc on mobile do animate but have no indication so it's easy to accidentally send multiples).

**Comment by t3chguy** [comment]
> Minor correction, gifs pasted into Element desktop don't animate. You need to manually save to device and manually upload.

That is misleading, thats a fault of your browser copying just a static PNG. See https://superuser.com/questions/1664216/why-does-google-chrome-and-safari-copy-image-as-png-representation. If you have a compatible browser or OS you can copy & paste GIFs into Element Desktop. e.g. this works from Finder on macOS.

**Comment by Destinyg133** [comment]
Still would be a lot easier if we could get direct integration iside client so we could easily search and send it

**Comment by hegdenischay** [comment]
Like I have [mentioned before](https://github.com/element-hq/element-meta/issues/321#issuecomment-2151434325), it's available as a widget for those who want it. 

TL; DR, use (stickerpicker)[https://github.com/maunium/stickerpicker] as a widget if you want Giphy, [replace some config options](https://view.matrix.org/room/!stickerpicker:maunium.net/?anchor=$QpB73g9d9WhM5fMMFF08c9ZepLBa0nj5RJYQHUpnkEg&offset=60) in it if you want Tenor, or use the [fork](https://github.com/pix/giphytenormatrixproxy) that @pix made if you want both. 

Additionally, there are a few matrix clients that have their own gif pickers, I can think of [Commet](https://commet.chat/) and [Gomuks web](https://github.com/tulir/gomuks) as options.

**Comment by Destinyg133** [comment]
> Like I have [mentioned before](https://github.com/element-hq/element-meta/issues/321#issuecomment-2151434325), it's available as a widget for those who want it.
> 
> TL; DR, use (stickerpicker)[https://github.com/maunium/stickerpicker] as a widget if you want Giphy, [replace some config options](https://view.matrix.org/room/!stickerpicker:maunium.net/?anchor=$QpB73g9d9WhM5fMMFF08c9ZepLBa0nj5RJYQHUpnkEg&offset=60) in it if you want Tenor, or use the [fork](https://github.com/pix/giphytenormatrixproxy) that [@pix](https://github.com/pix) made if you want both.
> 
> Additionally, there are a few matrix clients that have their own gif pickers, I can think of [Commet](https://commet.chat/) and [Gomuks web](https://github.com/tulir/gomuks) as options.

Ofcourse there are options and you are correct. But I think people want straight integration and official support rather then using stickerpicker and do devtools to include some kind of functionality. Commet is still heavy alfa. While it is cool that other clients try to show new functions in which matrix as protocol can be mainstream and better option then discord and services alike. Other clients still dont have support for element call afaik for example, so once again Element is only viable options for a lot of people.
Its great that devs show love for such a project but its without a doubt that some "fun" functions should come to Element without getting ignored for years. It can attract more people. And bigger project can only attract more people to help developers on open source projects.

**Comment by hegdenischay** [comment]
> > Like I have [mentioned before](https://github.com/element-hq/element-meta/issues/321#issuecomment-2151434325), it's available as a widget for those who want it.
> > TL; DR, use (stickerpicker)[https://github.com/maunium/stickerpicker] as a widget if you want Giphy, [replace some config options](https://view.matrix.org/room/!stickerpicker:maunium.net/?anchor=$QpB73g9d9WhM5fMMFF08c9ZepLBa0nj5RJYQHUpnkEg&offset=60) in it if you want Tenor, or use the [fork](https://github.com/pix/giphytenormatrixproxy) that [@pix](https://github.com/pix) made if you want both.
> > Additionally, there are a few matrix clients that have their own gif pickers, I can think of [Commet](https://commet.chat/) and [Gomuks web](https://github.com/tulir/gomuks) as options.
> 
> Ofcourse there are options and you are correct. But I think people want straight integration and official support rather then using stickerpicker and do devtools to include some kind of functionality. Commet is still heavy alfa. While it is cool that other clients try to show new functions in which matrix as protocol can be mainstream and better option then discord and services alike. Other clients still dont have support for element call afaik for example, so once again Element is only viable options for a lot of people. Its great that devs show love for such a project but its without a doubt that some "fun" functions should come to Element without getting ignored for years. It can attract more people. And bigger project can only attract more people to help developers on open source projects.

Seamless integration would be ideal and there have been PRs before that tried it and have failed for various reasons. But `stickerpicker` exists now, and while it's not ideal, it's definitely better than nothing. Also, there are [at least two](https://matrix.squirrel.rocks/stickers) sticker implementations competing for implementation, each with their pros and cons, so I am unsure if this issue will gain traction any time soon.

**Comment by moodler** [comment]
I wish Stickerpicker worked for me... 

We use our own home server.  I've got scalar.vector.im (the default) selected as the Integration Manager under Settings -> Security and Privacy.   However, trying to add a widget to any room just gives me 

"Cannot connect to integration manager
The integration manager is offline or it cannot reach your homeserver"

**Comment by t3chguy** [comment]
https://github.com/element-hq/element-integration-manager is the place to get help with Scalar @moodler

**Comment by Domoel** [comment]
so still no native integration for gifs?

**Comment by sigrdrifa** [comment]
Presumably there would be no matrix or backend changes required for this? A "frontend" picker that queries something like tenor or giphy and upon selecting, "copies" the GIF correctly into the message should suffice?

**Comment by Croydon** [comment]
> Presumably there would be no matrix or backend changes required for this? A "frontend" picker that queries something like tenor or giphy and upon selecting, "copies" the GIF correctly into the message should suffice?

That is surely possible today in clients. However, if you implement something like that nicely, you can save tons of resources (disk space, traffic) on both the server and clients.

**Comment by sigrdrifa** [comment]
> > Presumably there would be no matrix or backend changes required for this? A "frontend" picker that queries something like tenor or giphy and upon selecting, "copies" the GIF correctly into the message should suffice?
> 
> That is surely possible today in clients. However, if you implement something like that nicely, you can save tons of resources (disk space, traffic) on both the server and clients.

Yeah good point - I think it looks like a lot of attempts have been made but  no consensus has been agreed on for actual implementation. It looks like its not a technical problem, but more a feature / requirements definition problem

**Comment by MightyCreak** [comment]
I'm not sure the competition really cares about saving a few MB of gif when you can upload hundreds of MB of vids.

Also, I'd prefer that my client points to a GIF provider, but not my server. As a user I want to easily disable 3rd party GIF providers if I have privacy concerns.

This is also considering that, in any Matrix client right now, anyone can already upload a gif and use it in their messages.

**Comment by BrutalBirdie** [comment]
Day 3489, still sending GIF URLs occasionally in Element just to be reminded, this feature is still not present.
Emojis can't express the pain I feel not having this feature. ~:(

Common, I would really love if this would be added.
What does it need? Can one help somehow?

**Comment by Destinyg133** [comment]
Maybe we can use skmething like this? Comet integrated gif support 

https://github.com/commetchat/commet/tree/main/commet/lib/client/components/gif

**Comment by crabdancing** [comment]
"If ASI is achieved circa 2030, it will able to finally implement basic GIF picker support in Element" -- a techno-optimist, somewhere.

**Comment by yuceltoluyag** [comment]
<img width="600" height="788" alt="Image" src="https://github.com/user-attachments/assets/406742cb-c880-4dc3-9bb0-651d3efdfe1f" />

**Comment by iceHtwoO** [comment]
> Happy 10 year anniversary!!! 🎉🥳

I love getting email reminders like this once in a while

**Comment by dillfrescott** [comment]
Adding a comment so I can say I participated on this thread when my grandchildren ask why element still doesnt have a gif picker.

**Comment by MatzeG1t** [comment]
I think the underlying "non tech" issue is, that features of opensource products are usually prioritized by the implementing team. You need to regard another perspectives besides tech to make a product successful. I understand that initially products like Element et al are build because there is a user need of some people for better privacy, souverenity and detachment from big tech giants. This is great. In the coming product development the tech perspective is the dominant one, developing technological capable products as alternatives to Wassap, Telegram, Signal whatever.
If the product team of Element for example doesnt systematically incorporate the perspectives of the users besides the tech - the product will stick to a very small user group of techies and privacy focused users. That might be fine for the product team, but since this product team initially build the product to provide "a better alternative" for the society - it would make Element in this case even more successful if you systematically identify the basic uses cases of a messenger which just have to work so that users use the product.
GIF keyboard is just a basic feature of a messenger nowadays. Not having it available, just makes users reject the tool. It doesnt "feel right", "its not as good as my current messenger".

Sorry for the lenghty reply. But I'd really like to hear the answer of the developers or product responsibles to this. I cant imagine that they don't care. Maybe its just a point of "not knowing how to do it" - but this question can be answered :) Just one step by the other. Balancing between tech and users need would be really necessary. And most importantly not saying "I'm using it as a developer and my need is ..." . Listen to the fanbase and people who actually take the effort writing issues in github. Those are the early adopters and multiply the product usage...

**Comment by LordSousa** [comment]
Ain't no way this is still here.

**Comment by turt2live** [comment]
I believe the product team is well aware of the desire for this feature, so I'm locking this to avoid further comments which drive the conversation in circles.
